### PR TITLE
Fix function `has` with `LowCardinality` and `FixedString`

### LINKED ISSUE #64514: Lowcardinality: Logical error: 'this->n == rhs.n'
Report: https://s3.amazonaws.com/clickhouse-test-reports/0/2ba0d78eee6d6fc1c182c9c870e114a1

Minimized:

```
SET allow_suspicious_low_cardinality_types=1

CREATE TABLE fuzz (`ary` Array(LowCardinality(FixedString(4))),) ENGINE = MergeTree ORDER BY tuple();

INSERT INTO fuzz VALUES ([]) (['o','a']) (['e','a','b']);
INSERT INTO fuzz VALUES (['o','a','b','c']) (['e','a','b','c','d']);

SELECT count() FROM fuzz WHERE has(ary, materialize(toFixedString('a', 1)));
```

```
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007e1c3d74aeb3 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:78
#2  0x00007e1c3d6f2a30 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007e1c3d6da4c3 in __GI_abort () at abort.c:79
#4  0x0000000013e7465a in DB::abortOnFailedAssertion (description=...) at /mnt/ch/ClickHouse/src/Common/Exception.cpp:44
#5  0x000000001df9ec18 in DB::ColumnFixedString::compareAt (this=0x7e1b049bc300, p1=0, p2=0, rhs_=...) at /mnt/ch/ClickHouse/src/Columns/ColumnFixedString.h:135
#6  0x0000000012b10335 in DB::Impl::Main<DB::HasAction, true, unsigned long, unsigned long>::compare (left=..., right=..., i=0, j=1) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:107
#7  0x0000000012b0f6b3 in DB::Impl::Main<DB::HasAction, true, unsigned long, unsigned long>::process<1ul, DB::IColumn, DB::IColumn> (data=..., offsets=..., target=..., result=..., null_map_data=0x0, null_map_item=0x0) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:159
#8  0x0000000012a9d1b3 in DB::Impl::Main<DB::HasAction, true, unsigned long, unsigned long>::vector<DB::IColumn, DB::IColumn> (data=..., offsets=..., value=..., result=..., null_map_data=0x0, null_map_item=0x0) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:184
#9  0x0000000013877c0a in DB::FunctionArrayIndex<DB::HasAction, DB::NameHas>::dispatchConvertedLowCardinalityColumns (data=...) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:851
#10 0x0000000013876174 in DB::FunctionArrayIndex<DB::HasAction, DB::NameHas>::executeLowCardinality (arguments=...) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:836
#11 0x0000000013874fa4 in DB::FunctionArrayIndex<DB::HasAction, DB::NameHas>::executeOnNonNullable (this=0x7e19aa4aaaf8, arguments=..., result_type=...) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:585
#12 0x00000000138749cf in DB::FunctionArrayIndex<DB::HasAction, DB::NameHas>::executeArrayImpl (this=0x7e19aa4aaaf8, arguments=..., result_type=...) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:516
#13 0x0000000013874192 in DB::FunctionArrayIndex<DB::HasAction, DB::NameHas>::executeImpl (this=0x7e19aa4aaaf8, arguments=..., result_type=...) at /mnt/ch/ClickHouse/src/Functions/array/arrayIndex.h:451
#14 0x000000000bb3dd8d in DB::FunctionToExecutableFunctionAdaptor::executeImpl (this=0x7e19aa4aac60, arguments=..., result_type=..., input_rows_count=3) at /mnt/ch/ClickHouse/src/Functions/IFunctionAdaptors.h:21
#15 0x000000001a4ba149 in DB::IExecutableFunction::executeWithoutLowCardinalityColumns (this=0x7e19aa4aac60, args=..., result_type=..., input_rows_count=3, dry_run=false) at /mnt/ch/ClickHouse/src/Functions/IFunction.cpp:248
#16 0x000000001a4bb424 in DB::IExecutableFunction::executeWithoutSparseColumns (this=0x7e19aa4aac60, arguments=..., result_type=..., input_rows_count=3, dry_run=false) at /mnt/ch/ClickHouse/src/Functions/IFunction.cpp:307
#17 0x000000001a4bc4c4 in DB::IExecutableFunction::execute (this=0x7e19aa4aac60, arguments=..., result_type=..., input_rows_count=3, dry_run=false) at /mnt/ch/ClickHouse/src/Functions/IFunction.cpp:378
#18 0x000000001c49defc in DB::executeAction (action=..., execution_context=..., dry_run=false, allow_duplicates_in_input=false) at /mnt/ch/ClickHouse/src/Interpreters/ExpressionActions.cpp:616
#19 0x000000001c49d159 in DB::ExpressionActions::execute (this=0x7e19aa1df798, block=..., num_rows=@0x7e19a71ed498: 3, dry_run=false, allow_duplicates_in_input=false) at /mnt/ch/ClickHouse/src/Interpreters/ExpressionActions.cpp:750
#20 0x000000001c49ee0e in DB::ExpressionActions::execute (this=0x7e19aa1df798, block=..., dry_run=false, allow_duplicates_in_input=false) at /mnt/ch/ClickHouse/src/Interpreters/ExpressionActions.cpp:796
#21 0x000000001f0fff73 in DB::MergeTreeRangeReader::executePrewhereActionsAndFilterColumns (this=0x7e1b049bd030, result=...) at /mnt/ch/ClickHouse/src/Storages/MergeTree/MergeTreeRangeReader.cpp:1482
#22 0x000000001f0fe1ae in DB::MergeTreeRangeReader::read (this=0x7e1b049bd030, max_rows=3, ranges=...) at /mnt/ch/ClickHouse/src/Storages/MergeTree/MergeTreeRangeReader.cpp:1071
#23 0x000000001f136a61 in DB::MergeTreeReadTask::read (this=0x7e1b049bd000, params=...) at /mnt/ch/ClickHouse/src/Storages/MergeTree/MergeTreeReadTask.cpp:161
#24 0x0000000020205aec in DB::MergeTreeThreadSelectAlgorithm::readFromTask (this=0x7e1b078c64c0, task=..., params=...) at /mnt/ch/ClickHouse/src/Storages/MergeTree/MergeTreeSelectAlgorithms.h:38
#25 0x000000001f1300ed in DB::MergeTreeSelectProcessor::read (this=0x7e1b07956800) at /mnt/ch/ClickHouse/src/Storages/MergeTree/MergeTreeSelectProcessor.cpp:137
#26 0x0000000020266a9c in DB::MergeTreeSource::tryGenerate (this=0x7e1b078f7098) at /mnt/ch/ClickHouse/src/Storages/MergeTree/MergeTreeSource.cpp:226
#27 0x000000001f8835c3 in DB::ISource::work (this=0x7e1b078f7098) at /mnt/ch/ClickHouse/src/Processors/ISource.cpp:108
#28 0x000000001f8c6339 in DB::executeJob (node=0x7e19aa16ab00, read_progress_callback=0x7e1b079b91b0) at /mnt/ch/ClickHouse/src/Processors/Executors/ExecutionThreadContext.cpp:47
#29 0x000000001f8c6001 in DB::ExecutionThreadContext::executeTask (this=0x7e19b5cac100) at /mnt/ch/ClickHouse/src/Processors/Executors/ExecutionThreadContext.cpp:96
#30 0x000000001f8a499e in DB::PipelineExecutor::executeStepImpl (this=0x7e19aa10fdd8, thread_num=1, yield_flag=0x0) at /mnt/ch/ClickHouse/src/Processors/Executors/PipelineExecutor.cpp:272
#31 0x000000001f8a4dc4 in DB::PipelineExecutor::executeSingleThread (this=0x7e19aa10fdd8, thread_num=1) at /mnt/ch/ClickHouse/src/Processors/Executors/PipelineExecutor.cpp:238
#32 0x000000001f8a5ce7 in DB::PipelineExecutor::spawnThreads()::$_0::operator()() const (this=0x7e19b5c3d500) at /mnt/ch/ClickHouse/src/Processors/Executors/PipelineExecutor.cpp:372
#33 0x000000001f8a5c35 in std::__1::__invoke[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) (__f=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:394
#34 0x000000001f8a5c15 in std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) (__args=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:479
#35 0x000000001f8a5bf5 in std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>::operator()[abi:v15000]() (this=0x7e19b5c3d500) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:235
#36 0x000000001f8a5bc0 in std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()> >(std::__1::__function::__policy_storage const*) (__buf=0x7e19a71ef990)
    at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:716
#37 0x00000000125a6356 in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const (this=0x7e19a71ef990) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:848
#38 0x00000000125a5fb5 in std::__1::function<void ()>::operator()() const (this=0x7e19a71ef990) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:1187
#39 0x0000000013fc4872 in ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::worker (this=0x7e19b5cb1000, thread_it=...) at /mnt/ch/ClickHouse/src/Common/ThreadPool.cpp:462
#40 0x0000000013fcfdf1 in ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}::operator()() const (this=0x7e19a71efb50)
    at /mnt/ch/ClickHouse/src/Common/ThreadPool.cpp:219
#41 0x0000000013fcfdb5 in std::__1::__invoke[abi:v15000]<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&) (__f=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:394
#42 0x0000000013fcfd99 in std::__1::__apply_tuple_impl[abi:v15000]<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&, std::__1::tuple<>&>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&, std::__1::tuple<>&, std::__1::__tuple_indices<>) (__f=..., __t=...)
    at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/tuple:1789
#43 0x0000000013fcfbdd in std::__1::apply[abi:v15000]<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&, std::__1::tuple<>&>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&, std::__1::tuple<>&) (__f=..., __t=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/tuple:1798
#44 0x0000000013fcfa5f in ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&)::{lambda()#1}::operator()() (this=0x7e19b5c76740) at /mnt/ch/ClickHouse/src/Common/ThreadPool.h:251
#45 0x0000000013fcf8f5 in std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&)::{lambda()#1}&>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&) (__f=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:394
#46 0x0000000013fcf8d5 in std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&)::{lambda()#1}&>(ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&)::{lambda()#1}&) (__args=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:479
#47 0x0000000013fcf8b5 in std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&)::{lambda()#1}, void ()>::operator()[abi:v15000]() (this=0x7e19b5c76740)
    at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:235
#48 0x0000000013fcf880 in std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&)::{lambda()#1}, void ()> >(std::__1::__function::__policy_storage const*) (__buf=0x7e19a71f0150) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:716
#49 0x00000000125a6356 in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const (this=0x7e19a71f0150) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:848
#50 0x00000000125a5fb5 in std::__1::function<void ()>::operator()() const (this=0x7e19a71f0150) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/function.h:1187
#51 0x0000000013fc0d32 in ThreadPoolImpl<std::__1::thread>::worker (this=0x7e1c3c83b180, thread_it=...) at /mnt/ch/ClickHouse/src/Common/ThreadPool.cpp:462
#52 0x0000000013fc9411 in ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}::operator()() const (this=0x7e1a2a0ffe68) at /mnt/ch/ClickHouse/src/Common/ThreadPool.cpp:219
#53 0x0000000013fc93b5 in std::__1::__invoke[abi:v15000]<ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&) (__f=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__functional/invoke.h:394
#54 0x0000000013fc937d in std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, --Type <RET> for more, q to quit, c to continue without paging--q
Quit
(gdb) f 5
#5  0x000000001df9ec18 in DB::ColumnFixedString::compareAt (this=0x7e1b049bc300, p1=0, p2=0, rhs_=...) at /mnt/ch/ClickHouse/src/Columns/ColumnFixedString.h:135
135             chassert(this->n == rhs.n);
(gdb) p n
$1 = 4
(gdb) p rhs.n
$2 = 1
```

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed several cases in function `has` with non-constant `LowCardinality` arguments.

Fixes #64514.

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 22cfeb54fe9ea4b7dc9f1f458191b1edf965a5d9 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66088/22cfeb54fe9ea4b7dc9f1f458191b1edf965a5d9/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Functions/array/arrayIndex.h [hunk 1]
```diff
@@ -724,227 +688,153 @@ class FunctionArrayIndex : public IFunction
...
-            data.offsets, data.right,
-            data.result->getData(),
-            data.maps.first, data.maps.second);
+        auto arguments_copy = arguments;
+        arguments_copy[0].column = std::move(array_column);
+        arguments_copy[0].type = std::move(array_type);
+        arguments_copy[0].name = arguments[0].name;
 
-        data.moveResult();
-        return true;
```
Comment (CurtizJ) [inline src/Functions/array/arrayIndex.h:858]: The problem was here: we executed "generic" implementation without cast of columns to common type. Also this code looked strange.

File: src/Functions/array/arrayIndex.h [hunk 2]
```diff
@@ -635,39 +614,33 @@ class FunctionArrayIndex : public IFunction
...
     static ColumnPtr executeIntegral(const ColumnsWithTypeAndName & arguments)
     {
-        const ColumnArray * const left = checkAndGetColumn<ColumnArray>(arguments[0].column.get());
-
-        if (!left)
+        const auto * array = checkAndGetColumn<ColumnArray>(arguments[0].column.get());
+        if (!array)
             return nullptr;
 
-        const ColumnPtr right_converted_ptr = arguments[1].column->convertToFullColumnIfLowCardinality();
```
Comment (CurtizJ) [inline src/Functions/array/arrayIndex.h:643]: Another issue, that we could convert column to full several times: in `executeIntegral`, `executeString`...

**Review by alexey-milovidov — APPROVED** [review]

**Comment by CurtizJ** [comment]
Bugfix validation: #65946.
clang-tidy build: #66161.
02956_rocksdb_bulk_sink and test_host_regexp_multiple_ptr_v4 are flaky according to the CI database.

**Pull request merged**